### PR TITLE
Build in-memory etcd model for deterministic simulation

### PR DESCRIPTION
**Pull request by ahrav**
Introduce `SimulatedEtcdKV` — a feature-gated in-process implementation of
the etcd KV subset (point reads, prefix scans, CAS transactions, lease semantics) used by the coordination backend. Enables deterministic simulation without a live etcd server.

Includes `SimEtcdFaultConfig` for seeded fault injection and comprehensive
test coverage of KV operations, revision metadata tracking, and lease expiry logic.

## Summary by CodeRabbit

* **New Features**
  * Added a feature-gated, deterministic in-memory etcd KV simulator for testing (reads, prefix scans, atomic transactions, leases, time control, configurable fault injection) and introduced additional public types for findings/migration reporting and observability.

* **Tests**
  * Expanded testing and verification infrastructure with concurrency, fuzzing, and broader simulation/invariant coverage across components.

* **Refactor**
  * Reduced external visibility of a validation helper to limit public surface.

* **Documentation**
  * Updated coordination docs to describe the new simulation model and testing capabilities.

### CODE REVIEW

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 1]
```diff
@@ -0,0 +1,1056 @@
...
+            let deleted_any = match self.remove_lease_and_keys(lease_id) {
+                Ok(deleted_any) => deleted_any,
+                Err(SimEtcdError::LeaseNotFound { .. }) => false,
+                Err(err) => panic!("internal lease expiry failed: {err}"),
+            };
+            if deleted_any {
+                self.revision = self.revision.saturating_add(1);
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: Investigated this and wrote a test: granted 3 leases (all TTL=5), attached one key to each, then expired all 3 at once via `tick(5)`. The test confirmed the simulator produces 3 separate revision bumps (one per expired lease with keys), which is the correct behavior.

etcd's Lessor calls `Revoke()` individually for each expired lease via `revokeExpiredLease()`. Each revocation is a separate transaction that independently bumps the revision. Three leases expiring at the same tick produce three revision increments, not one. The simulator's per-lease revision bump matches real etcd.

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+            if !staged.touched_keys.insert(key.clone()) {
+                return Err(SimEtcdError::DuplicateMutation { key: key.clone() });
+            }
+            staged.deleted_keys.insert(key.clone());
+        }
+        if !staged.deleted_keys.is_empty() {
+            staged.changed = true;
+        }
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: Investigated this — the issue is real in principle but harmless in the current code, as you acknowledged. In all traced scenarios the `changed` flag ends up correct because it's a boolean and re-setting `true` to `true` is a no-op.

This is now moot: the `StagedTxn` redesign (for the txn op ordering fix) eliminated the `deleted_keys` set entirely. The new `stage_delete` tracks matched keys in a local `BTreeSet` and sets `changed = true` only if `!keys_to_delete.is_empty()` — scoped to the current op, not cumulative state.

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 3]
```diff
@@ -0,0 +1,1193 @@
...
+#[derive(Debug, Default)]
+struct StagedTxn {
+    ops: Vec<StagedOp>,
+    /// Keys staged for put that have not yet been cancelled by a later delete.
+    /// Used so that `stage_delete` can resolve against previously staged puts
+    /// in addition to the pre-transaction store.
+    pending_puts: BTreeSet<Vec<u8>>,
+    /// Keys already staged for put (including those later deleted).
+    /// Guards against two puts to the same key in the same transaction.
+    all_put_keys: BTreeSet<Vec<u8>>,
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: Reasonable suggestion — applied. Added a doc note to `StagedTxn` explaining that `all_put_keys` is never cleared, so `put(k) → delete(k) → put(k)` is rejected with `DuplicateMutation`, and that this simplification is acceptable because the coordination backend never issues such patterns.

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 4]
```diff
@@ -0,0 +1,1510 @@
...
+        self.next_lease_id = self
+            .next_lease_id
+            .checked_add(1)
+            .ok_or(SimEtcdError::LeaseIdExhausted)?;
+
+        let ttl_u64 = u64::try_from(ttl).expect("TTL validated positive");
+        let expires_at = self
+            .now
+            .checked_add(ttl_u64)
+            .expect("lease expiry overflow");
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: Not applying this change. These panics are appropriate for their context:

1. **`u64::try_from(ttl).expect("TTL validated positive")`** — provably unreachable; the `ttl <= 0` guard 3 lines above guarantees `ttl > 0`, making the conversion infallible.

2. **`checked_add(...).expect("lease expiry overflow")` / `tick` overflow** — `now` starts at 0 and advances by simulation ticks. Reaching `u64::MAX` (18.4 quintillion) would require more ticks than any simulation will ever run. This is the same class as "usize overflow on a 64-bit machine" — technically possible, practically unreachable.

3. **`bump_revision` overflow** — same reasoning. `i64::MAX` revisions is ~9.2 quintillion. No simulation reaches this.

4. **`validate_ppm` assert** — construction-time invariant for test setup. Passing `ppm > 1_000_000` is a programming error, not a runtime edge case. `assert!` is the idiomatic Rust response.

Converting these to `Result` would propagate error handling to every `tick()`, `lease_grant()`, and fault-config call site for conditions that cannot occur in practice. The complexity cost outweighs the theoretical safety gain, especially in test-support infrastructure that only runs under `cfg(test)` or `feature = "test-support"`.

File: crates/gossip-coordination-etcd/src/sim_etcd_kv.rs [hunk 5]
```diff
@@ -0,0 +1,1555 @@
...
+                        }
+                        self.delete_key(&key);
+                    }
+                    responses.push(PbResponseOp {
+                        response: Some(PbTxnOpResponse::ResponseDeleteRange(PbDeleteResponse {
+                            header,
+                            deleted,
+                            prev_kvs: Vec::new(),
+                        })),
+                    });
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-coordination-etcd/src/sim_etcd_kv.rs]: Investigated this and wrote a test (`multi_op_txn_produces_header_per_response`) with a 3-op txn (three puts). The test compiles and passes — `PbResponseHeader` is `Copy` in the etcd-client version used by this project (all fields are `u64`/`i64` primitives, and the prost codegen for this crate includes `Copy`). The `Option<PbResponseHeader>` binding is copied on each use, not moved.

The web query result about prost's default derives is correct in general, but this project's etcd-client dependency generates `Copy` for `PbResponseHeader` specifically. No change needed.
